compiler/rustc_hir_typeck/src/expectation.rs

```diff
@@ -46,7 +46,11 @@ impl<'a, 'tcx> Expectation<'tcx> {
         match *self {
             ExpectHasType(ety) => {
                 let ety = fcx.shallow_resolve(ety);
-                if !ety.is_ty_var() { ExpectHasType(ety) } else { NoExpectation }
+                if matches!(ety.kind(), ty::Infer(ty::TyVar(_)) | ty::Alias(ty::Opaque, _)) {
```

Comment:
Conceptually, opaque aliases are kinda the same as type vars -- we don't really know enough about the type to use it as an expectation, so just throw the expectation away.

Reply:
do we instead want `ety.is_known_rigid()` here? :thinking:

Reply:
we don't want to lose projections (or weak type aliases). But having a helper method on `Ty` that checks for `opaque | var` seems good.

Reply:
this may cause issues for us when trying to remove `replace_opaque_types_with_inference_vars`. in the old solver we could go: opaque -> infer -> partially constrained -> keep as expectation, while the new solver would go: opaque -> stays as opaque -> partially constrained hidden type -> drop expectation

it might be nicer to change `try_structurally_resolve_type` to normalize opaque types if they normalize to something whicih is not an infer var.